storage: TestRaftBlockedReplica is flaky (by design)

That test verifies that with one blocked replica, we still process Raft messages for the others. However, there is a limited worker pool, and with enough ticks (or other activity), all workers will be stuck on the blocked replica.

This is happening frequently with `TAGS=deadlock` (presumably due to the altered timings in this test) and can be exacerbated by  setting `storeSchedulerConcurrency=1`.

Fails on `master`. I haven't investigated exactly why, but it doesn't happen without the `deadlock` build tag (at least not as easily). This is on my 32cpu slave and I see 64 worker threads (I expected 2*32*3(=num_nodes), but that's not what's coming out). Doesn't seem that they're all blocked. I'll do some more digging but I can (and would prefer to) hand this off.

```
make stress PKG=./storage TAGS=deadlock TESTS=TestRaftBlockedReplica TESTTIMEOUT=5m

[wait 31s, which is the timeout after which `deadlock` triggers early abort]
POTENTIAL DEADLOCK:
Previous place where the lock was grabbed
goroutine 334 lock 0xc420204368
replica.go:2011 storage.(*Replica).tick { r.raftMu.Lock() } <<<<<
store.go:2853 storage.(*Store).processTick { if exists, err = r.tick(); err != nil { }
scheduler.go:188 storage.(*raftScheduler).worker { if s.processor.processTick(id) { }
scheduler.go:160 storage.(*raftScheduler).Start.func1 { s.worker(stopper) }
../util/stop/stopper.go:188 stop.(*Stopper).RunWorker.func1 { f() }

Have been trying to lock it again for more than 30s
goroutine 334 lock 0xc420204368
replica.go:2011 storage.(*Replica).tick { r.raftMu.Lock() } <<<<<
store.go:2853 storage.(*Store).processTick { if exists, err = r.tick(); err != nil { }
scheduler.go:188 storage.(*raftScheduler).worker { if s.processor.processTick(id) { }
scheduler.go:160 storage.(*raftScheduler).Start.func1 { s.worker(stopper) }
../util/stop/stopper.go:188 stop.(*Stopper).RunWorker.func1 { f() }


Here is what goroutine 334 doing now
goroutine 334 [running]:
github.com/sasha-s/go-deadlock.stacks(0x20ff760, 0xc420052018, 0x0)
       	/home/tschottdorf/src/github.com/sasha-s/go-deadlock/stacktraces.go:101 +0x79
github.com/sasha-s/go-deadlock.lock(0xc420b21b20, 0x15f4060, 0xc420204368)
       	/home/tschottdorf/src/github.com/sasha-s/go-deadlock/deadlock.go:179 +0x767
github.com/sasha-s/go-deadlock.(*Mutex).Lock(0xc420204368)
       	/home/tschottdorf/src/github.com/sasha-s/go-deadlock/deadlock.go:55 +0x78
github.com/cockroachdb/cockroach/storage.(*Replica).tick(0xc420204300, 0xc42021c900, 0x0, 0x0)
       	/home/tschottdorf/src/github.com/cockroachdb/cockroach/storage/replica.go:2011 +0x51
github.com/cockroachdb/cockroach/storage.(*Store).processTick(0xc420197400, 0x2, 0xc42066eea0)
       	/home/tschottdorf/src/github.com/cockroachdb/cockroach/storage/store.go:2853 +0xe7
github.com/cockroachdb/cockroach/storage.(*raftScheduler).worker(0xc42019d340, 0xc4201b61b0)
       	/home/tschottdorf/src/github.com/cockroachdb/cockroach/storage/scheduler.go:188 +0x35f
github.com/cockroachdb/cockroach/storage.(*raftScheduler).Start.func1()
       	/home/tschottdorf/src/github.com/cockroachdb/cockroach/storage/scheduler.go:160 +0x33
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc4201b61b0, 0xc4200f77c0)
       	/home/tschottdorf/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:188 +0x7d
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
       	/home/tschottdorf/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:189 +0x66

Other goroutines holding locks:
goroutine 1624 lock 0xc4202c2aa0
scheduler.go:263 storage.(*raftScheduler).enqueueN { s.mu.Lock() } <<<<<
scheduler.go:289 storage.(*raftScheduler).EnqueueRaftTick { s.signal(s.enqueueN(stateRaftTick, ids...)) }
store.go:2891 storage.(*Store).raftTickLoop.func1 { s.scheduler.EnqueueRaftTick(rangeIDs...) }
../util/stop/stopper.go:188 stop.(*Stopper).RunWorker.func1 { f() }

goroutine 912 lock 0xc420638068
replica.go:1771 storage.(*Replica).handleRaftReady { r.raftMu.Lock() } <<<<<
store.go:2817 storage.(*Store).processReady { if err := r.handleRaftReady(IncomingSnapshot{}); err != nil { }
scheduler.go:193 storage.(*raftScheduler).worker { s.processor.processReady(id) }
scheduler.go:160 storage.(*raftScheduler).Start.func1 { s.worker(stopper) }
../util/stop/stopper.go:188 stop.(*Stopper).RunWorker.func1 { f() }

goroutine 912 lock 0xc420638070
replica.go:461 storage.(*Replica).withRaftGroup { r.mu.Lock() } <<<<<
replica.go:2005 storage.(*Replica).handleRaftReadyRaftMuLocked { }) }
replica.go:1773 storage.(*Replica).handleRaftReady { return r.handleRaftReadyRaftMuLocked(inSnap) }
store.go:2817 storage.(*Store).processReady { if err := r.handleRaftReady(IncomingSnapshot{}); err != nil { }
scheduler.go:193 storage.(*raftScheduler).worker { s.processor.processReady(id) }
scheduler.go:160 storage.(*raftScheduler).Start.func1 { s.worker(stopper) }
../util/stop/stopper.go:188 stop.(*Stopper).RunWorker.func1 { f() }
ERROR: exit status 2

193 runs completed, 1 failures, over 32s
FAIL
Makefile:146: recipe for target 'stress' failed
make: *** [stress] Error 1

```

--- a/storage/client_raft_test.go
+++ b/storage/client_raft_test.go
@@ -2613,7 +2613,20 @@ func TestRaftBlockedReplica(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	rep.RaftLock()
+
+	// NB: We perform the actual locking on a different goroutine in order to
+	// workaround a spurious inconsistent lock order warning when running with
+	// TAGS=deadlock. The issue is that we're grabbing Replica 2's raftMu and
+	// then later Replica 1's from the same goroutine due to the direct calling
+	// of client.SendWrapped down the callstack into the Replica code (via the
+	// local RPC optimization).
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		rep.RaftLock()
+		wg.Done()
+	}()
+	wg.Wait()
 	defer rep.RaftUnlock()
 
 	// Verify that we're still ticking the non-blocked replica.

